src/lib.rs

```diff
@@ -51,144 +49,406 @@
+/// [Reference](https://developer.spotify.com/documentation/general/guides/authorization-guide/)
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Token {
+    /// An access token that can be provided in subsequent calls
+    pub access_token: String,
+    /// The time period for which the access token is valid.
+    #[serde(with = "duration_second")]
+    pub expires_in: Duration,
+    /// The valid time for which the access token is available represented
+    /// in ISO 8601 combined date and time.
+    pub expires_at: Option<DateTime<Utc>>,
+    /// A token that can be sent to the Spotify Accounts service
+    /// in place of an authorization code
+    pub refresh_token: Option<String>,
+    /// A list of [scopes](https://developer.spotify.com/documentation/general/guides/scopes/)
+    /// which have been granted for this `access_token`
+    /// You could use macro [scopes!](crate::scopes) to build it at compile time easily
+    #[serde(default, with = "space_separated_scope")]
+    pub scope: HashSet<String>,
+}
+
+impl Default for Token {
+    fn default() -> Self {
+        Token {
+            access_token: String::new(),
+            expires_in: Duration::seconds(0),
+            expires_at: Some(Utc::now()),
+            refresh_token: None,
+            scope: HashSet::new(),
+        }
+    }
+}
+
+impl Token {
+    /// Tries to initialize the token from a cache file.
+    // TODO: maybe ClientResult for these things instead?
+    pub fn from_cache<T: AsRef<Path>>(path: T) -> Option<Self> {
+        let mut file = fs::File::open(path).ok()?;
+        let mut tok_str = String::new();
+        file.read_to_string(&mut tok_str).ok()?;
+
+        serde_json::from_str::<Token>(&tok_str).ok()
+    }
+
+    /// Saves the token information into its cache file.
+    pub fn write_cache<T: AsRef<Path>>(&self, path: T) -> ClientResult<()> {
+        let token_info = serde_json::to_string(&self)?;
+
+        let mut file = fs::OpenOptions::new().write(true).create(true).open(path)?;
+        file.set_len(0)?;
+        file.write_all(token_info.as_bytes())?;
+
+        Ok(())
+    }
+
+    /// Check if the token is expired
+    pub fn is_expired(&self) -> bool {
+        self.expires_at
+            .map_or(true, |x| Utc::now().timestamp() > x.timestamp())
+    }
+}
+
+/// Simple client credentials object for Spotify.
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct Credentials {
+    pub id: String,
+    pub secret: String,
+}
+
+impl Credentials {
+    pub fn new(id: &str, secret: &str) -> Self {
+        Credentials {
+            id: id.to_owned(),
+            secret: secret.to_owned(),
+        }
+    }
+
+    /// Parses the credentials from the environment variables
+    /// `RSPOTIFY_CLIENT_ID` and `RSPOTIFY_CLIENT_SECRET`. You can optionally
+    /// activate the `env-file` feature in order to read these variables from
+    /// a `.env` file.
+    pub fn from_env() -> Option<Self> {
```

Comment:
Perhaps we could return `ClientResult` for `from_env()` functions instead of `Option<Self>`?

Reply:
The thing is that `env::var` returns just an `Option`, so there's nothing else to report with a `Result` in that function.

I also wanted to do this for `Token::from_cache`, but then I would have to re-think this in `src/client_creds.rs`:

```rust
    /// Tries to read the cache file's token, which may not exist.
    ///
    /// Similarly to [`Self::write_token_cache`], this will already check if the
    /// cached token is enabled and return `None` in case it isn't.
    #[maybe_async]
    pub async fn read_token_cache(&self) -> Option<Token> {
        if !self.get_config().token_cached {
            return None;
        }

        let token = Token::from_cache(&self.get_config().cache_path)?;
        if token.is_expired() {
            // Invalid token, since it doesn't have at least the currently
            // required scopes or it's expired.
            // TODO: update with error?
            None
        } else {
            Some(token)
        }
    }
```

In that case we now have to return an error when the scopes don't match instead of (`Some`/`None`), and I'm not sure which one to pick. Perhaps we should add a new variant?

Reply:
Well, I think `Some/None` is more reasonable, since we should return `None` when we could not find matched data, which is following the semantics of `Option` type.